src/wallet/mod.rs

```diff
@@ -750,6 +750,20 @@ where
 
         fee_amount += fee_rate.fee_wu(tx.weight());
 
+        // Segwit transactions' header is 2WU larger than legacy txs' header,
+        // as they contain a witness marker (1WU) and a witness flag (1WU) (see BIP144).
+        // At this point we really don't know if the resulting transaction will be segwit
+        // or legacy, so we just add this 2WU to the fee_amount - overshooting the fee amount
+        // is better than undershooting it.
+        // If we pass a fee_amount that is slightly higher than the final fee_amount, we
+        // end up with a transaction with a slightly higher fee rate than the requested one.
+        // If, instead, we undershoot, we may end up with a feerate lower than the requested one
+        // - we might come up with non broadcastable txs!
+        fee_amount += fee_rate.fee_wu(2);
+        // FIXME: we should apply this same undershooting/overshooting heuristic to the drain output
+        // as well: when we add a drain_output that we didn't consider previously in the
+        // fee_amount, we're lowering the transaction fee rate.
+
```

Comment:
I don't understand this: As this is done before coin selection, the selected amount will be considering the extra fees derived of these 2 WUs as part of the amount to pay, so any excess (dropped as fees or used as change) will be produced based on the same extra fee.
How would the `drain_output` reduce the transaction fee rate?

Reply:
Didn't we say that it should be the coin selection to decide whether to add the drain output or not? I think we've been talking about this forever, the current logic of checking the remainder after the coin selection is done is suboptimal, so we should just give the coin selection the weight cost of the drain output and it will decide whether to add it or not.

Which should remove this problem completely if I understand it correctly

Reply:
> As this is done before coin selection, the selected amount will be considering the extra fees derived of these 2 WUs as part of the amount to pay, so any excess (dropped as fees or used as change) will be produced based on the same extra fee.
How would the drain_output reduce the transaction fee rate?

Yeah, you're right, at this point I'm not even sure anymore. I'm positive that sometimes we return a tx with a lower fee rate, but I can't figure out why.
Anywoo, I'll remove the comment as it's obviously wrong, and I'll keep working on this for another PR :)

>  I think we've been talking about this forever

Eh, review #630 then...